src/java/main/org/apache/zookeeper/server/NettyServerCnxnFactory.java

```diff
@@ -109,6 +110,20 @@ public void channelConnected(ChannelHandlerContext ctx,
                     zkServer, NettyServerCnxnFactory.this);
             ctx.setAttachment(cnxn);
 
+            InetAddress addr = ((InetSocketAddress)cnxn.channel.getRemoteAddress()).getAddress();
+            Set<NettyServerCnxn> s = ipMap.get(addr);
+            if (s != null) {
+                synchronized (s) {
+                    int cnxncount = s.size();
+                    if (maxClientCnxns > 0 && cnxncount >= maxClientCnxns) {
+                        LOG.warn("Closed new connection from address {} ({} connections established) - max is {}",
+                                addr, cnxncount, maxClientCnxns);
+                        cnxn.close();
```

Comment:
Could you double-check that this is the right order of closing cnxn and channel, please?